proposals/0000-dropped-assets-claimers.md

```diff
@@ -0,0 +1,139 @@
+---
+Title: Dropped Assets Claimers
+Number: 0
+Status: Draft
+Version: 0
+Authors:
+ - Daniel Shiposha
+Created: 2023-06-21
+Impact: Low
+Requires:
+Replaces:
+---
+
+## Summary
+
+The proposed change provides a way to set claimers to potentially dropped assets.
+Currently, determining a claimer of dropped assets is implementation-specific, and there is no way to set a custom claimer.
+The ability to set a custom claimer origin makes it easier to rescue the dropped assets, especially in the case of a reserve-based transfer.
+
+## Motivation
+
+XCM is an important feature of the Polkadot ecosystem that enables inter-chain communication and value transfer between ecosystem parachains. When assets (fungible or non-fungible) are transferred using XCM, they may be dropped due to miscellaneous errors during the execution. The probability of such an event is low, but it is not negligible, and dropped assets will need to be recovered to restore the user control of them. The process of such recovery is called "claiming", and the account that has the permission to perform it is called "claimer".
+
+Currently, determining a claimer of dropped assets is implementation-specific, and there is no way to set a custom claimer. The implementation could choose a predictable claimer origin such as the same origin as the origin of the XCM message. But also, this choice could be arbitrary if there is no origin, e.g., if `ClearOrigin` was executed before an error happened.
+
+The latter case includes reserve-based transfers. Given that parachains mostly rely on reserve-based transfers, a way to define a custom claimer for the dropped assets makes rescuing more convenient since it provides more ways to do so compared to an arbitrary claimer setting defined by a specific implementation.
+
+Having this could become especially important when transferring multiple currencies or NFTs:
+ * When transferring multiple currencies, one must choose a currency to pay execution fees.
+   If there are insufficient funds in the chosen currency, **all** the assets will be dropped. And the dropped amount could be significant, so we need a convenient way to recover it.
+ * When transferring NFTs, we must transfer some fungibles alongside them to pay fees. This is precisely the same situation as with multiple currencies. However, an NFT representing a unique object could be subjectively very valuable to a user, so its dropping can be compared to dropping a large number of fungibles.
+
+This RFC proposes adding a new instruction, `AddDropedAssetsClaimer`. The new instruction sets a custom claimer to particular dropped assets defined by an asset filter.
+
+## Specification
+
+The instruction would have the following specification:
+
+```rust
+AddDropedAssetsClaimer { claimer: MultiLocation, assets: MultiAssetFilter }
```

Comment:
Hmm, your question is more complex than I thought...
What if we had not 10 fungibles dropped assets in total but, say, 8?
And one claimer can claim 7, and the other can claim 3. What should we do in this situation?

Give me some time to think, please :)

Reply:
It seems the Definite filter should define "the maximum amount of a claimable asset".

Considering the last example:
* Total amount: 8 fungibles
* Claimer A can claim **up to** 7 fungibles
* Claimer B can claim **up to** 3 fungibles

So, if claimer B first claims their 3 fungibles, only `8 - 3 = 5` fungibles will be left for A to claim.
I don't see a better way. However, this doesn't look good either.

Maybe, as Gavin pointed out in another comment, there shouldn't be asset splitting. I thought this would make the setting of claimants more general, but I can't see a use case for this yet. It may not be needed now.

Also, I'm not sure about the list of claimants... E.g., in a transfer program, one could set the sender and the beneficiary as claimants to the assets. But is that useful? Maybe setting one of them is enough?

I can see that implementing a list of claimants might be complex (considering pallet-xcm, not the executor), but at the same time implementing just one claimer will be trivial.